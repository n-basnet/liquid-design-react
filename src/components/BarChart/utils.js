import uniqid from 'uniqid'
import { find, pluck, head, prop, uniqBy } from 'ramda'

import { DEFAULT_COLORS } from '~/components/BarChart/consts'
import { times, getElementModular } from '~/utils/misc'
import { SCREEN_SIZES } from '~/utils/styling'

const getLabels = data =>
  data
    .reduce((labels, { values }) => {
      const valuesLabels = pluck('label', values)
      return uniqBy(prop('name'), [...labels, ...valuesLabels])
    }, [])
    .map((label, i) => ({
      id: `l-${uniqid()}`,
      color: getElementModular(i, DEFAULT_COLORS),
      ...label,
    }))

const getLargestValue = data => {
  const allValues = pluck('values', data).map(pluck('value'))
  return head(allValues.reduce((a, b) => [...a, ...b], []).sort((a, b) => b - a))
}

const findByName = (name, array) => find(element => element.name === name, array)

const roundToThousands = num => Math.round(num / 1000) * 1000
export const getTicks = (data, offset = 0, amount = 6) => {
  if (data.length === 0) {
    return []
  }
  const maximum = (100 * getLargestValue(data)) / (100 - offset)
  const tickValue = roundToThousands(maximum / amount)
  return times(amount + 1).map(i => i * tickValue)
}

export const recomputeData = data => {
  const largestValue = getLargestValue(data)
  const labels = getLabels(data)

  return {
    data: data.map((point, i) => ({
      ...point,
      id: `p-${uniqid()}`,
      values: point.values.map(value => ({
        id: `v-${uniqid()}`,
        valueAsPercentage: (value.value * 100) / largestValue,
        ...value,
        label: findByName(value.label.name, labels),
      })),
    })),
    labels,
  }
}

export const isMobile = () => window.innerWidth <= SCREEN_SIZES.phone
