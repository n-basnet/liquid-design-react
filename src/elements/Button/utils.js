import { path } from 'ramda'

export const bgColorSelector = (props, shade) => {
  const colorsMap = {
    primary: 'primary',
    secondary: 'sensitiveGrey',
    highlight: 'vibrantYellow',
  }
  return props.theme.colors[colorsMap[props.appearance]][shade]
}

export const colorSelector = props => {
  const colorsMap = {
    primary: ['white', 'base'],
    highlight: ['richBlack', 'base'],
    secondary: props.disabled ? ['sensitiveGrey', 'darker'] : ['primary', 'base'],
  }
  return path(colorsMap[props.appearance], props.theme.colors)
}

export const opacitySelector = props => {
  const opacityMap = {
    primary: '0.5',
    secondary: '1',
    highlight: '0.3',
  }

  return opacityMap[props.appearance]
}

export const iconOpacitySelector = props => {
  const opacityMap = {
    primary: '0.5',
    secondary: '0.3',
    highlight: '0.3',
  }

  return opacityMap[props.appearance]
}
