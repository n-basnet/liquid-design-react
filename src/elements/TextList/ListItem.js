import React from 'react'
import PropTypes from 'prop-types'

const ListItem = ({ name, children }) => (
  <li>
    {name}
    {children}
  </li>
)

ListItem.propTypes = {
  children: PropTypes.node,
  name: PropTypes.string.isRequired,
}

ListItem.defaultProps = {
  children: null,
}

export default ListItem
