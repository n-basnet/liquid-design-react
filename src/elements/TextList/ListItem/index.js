import React from 'react'
import PropTypes from 'prop-types'

export const ListItem = ({ name, children }) => (
  <li>
    {name}
    {children}
  </li>
)

ListItem.propTypes = {
  children: PropTypes.node,
  name: PropTypes.string.isRequired,
}

ListItem.defaultProps = {
  children: null,
}
