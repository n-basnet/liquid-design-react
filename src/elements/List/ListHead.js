import styled, { css } from 'styled-components'

import { ICON_CLASSNAME } from '~/elements/Icon'

export const ListHead = styled.div`
  display: flex;
  max-width: 100%;
  min-height: 50px;
  padding: 11px 15px;
  border-radius: 6px 6px 0 0;
  font-size: 16px;
  line-height: 1.75;
  ${props => css`
    border-bottom: 1px solid ${props.theme.colors.sensitiveGrey.darker};
    font-weight: ${props.theme.fontWeight.black};
    ${props.grey &&
      css`
        background-color: ${props.theme.colors.sensitiveGrey.base};
      `};
    .${ICON_CLASSNAME} {
      margin-right: 10px;
      margin-top: 6px;
      min-width: 16px;
    }
  `};
`
