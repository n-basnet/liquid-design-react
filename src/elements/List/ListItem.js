import styled, { css } from 'styled-components'

import { ICON_CLASSNAME } from '~/elements/Icon'
import { cursorValue, nonTouchDevicesHoverStyles } from '~/utils/styling'

export const ListItem = styled.li`
  display: flex;
  max-width: 100%;
  min-height: 50px;
  padding: 12px 15px;
  list-style: none;
  font-size: 14px;
  line-height: 1.75;
  :last-child {
    border-radius: 0 0 6px 6px;
    border: none;
  }
  ${props => css`
    border-bottom: 1px solid ${props.theme.colors.sensitiveGrey.darker};
    ${cursorValue}
    ${!props.disabled &&
      props.onClick &&
      nonTouchDevicesHoverStyles(`
      background-color: ${
  props.grey ? props.theme.colors.sensitiveGrey.dark : props.theme.colors.white
};
      color: ${props.theme.colors.primary.base};
      font-weight: ${props.theme.fontWeight.black};
      transition: ${props.theme.transition};
      .${ICON_CLASSNAME} {
        svg {
          fill: ${props.theme.colors.primary.base};
        }
      }
      cursor: pointer;
    `)}
    ${props.active &&
      css`
        font-weight: ${props.theme.fontWeight.black};
        color: ${props.theme.colors.primary.base};
      `};
    ${props.grey &&
      css`
        background-color: ${props.theme.colors.sensitiveGrey.base};
      `};
    ${props.disabled &&
      css`
        span,
        svg {
          opacity: 0.3;
        }
      `};
    .${ICON_CLASSNAME} {
      margin-right: 10px;
      margin-top: 4px;
      min-width: 16px;
      ${props =>
    props.active &&
        css`
          svg {
            fill: ${props.theme.colors.primary.base};
          }
        `};
  `};
`
