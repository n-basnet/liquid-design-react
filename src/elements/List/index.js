import React from 'react'
import PropTypes from 'prop-types'

import Icon from '~/elements/Icon'
import { ListHead } from '~/elements/List/ListHead'
import { ListItem } from '~/elements/List/ListItem'

const List = ({
  activeItemIndex,
  disabledItemIndex,
  grey,
  icon,
  items = [],
  listHead,
  onClick,
}) => {
  const ListItems = items.map((item, index) => (
    <ListItem
      active={index === activeItemIndex}
      disabled={index === disabledItemIndex}
      grey={grey}
      icon={icon}
      key={index}
      onClick={index !== disabledItemIndex ? () => onClick(index) : undefined}
    >
      {icon && (
        <Icon
          color={activeItemIndex ? 'primary' : 'richBlack'}
          name={icon}
          size={16}
        />
      )}
      <span>{item}</span>
    </ListItem>
  ))

  return (
    <ul>
      <ListHead grey={grey}>
        {icon && (
          <Icon
            color={activeItemIndex ? 'primary' : 'richBlack'}
            name={icon}
            size={16}
          />
        )}
        {listHead}
      </ListHead>
      {ListItems}
    </ul>
  )
}

List.propTypes = {
  activeItemIndex: PropTypes.number,
  disabledItemIndex: PropTypes.number,
  grey: PropTypes.bool,
  icon: PropTypes.string,
  items: PropTypes.array,
  listHead: PropTypes.string,
  onClick: PropTypes.func,
}

List.defaultProps = {
  activeItemIndex: null,
  disabledItemIndex: null,
  grey: false,
  icon: null,
  items: PropTypes.array,
  listHead: null,
  onClick: null,
}

export default List
