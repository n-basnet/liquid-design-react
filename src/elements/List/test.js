import List from '.'
import { ListItem } from '../../elements/List/ListItem'
import { everyComponentTestSuite, getWrapper } from '../../utils/testUtils'

const items = [
  { name: 'item 1', onClick: jest.fn() },
  { name: 'item 2', isDisabled: true },
  { name: 'item 3' },
]

describe('List', () => {
  const defaultProps = {
    grey: true,
    items,
    listHead: { name: 'List Head' },
    activeItemIndex: 1,
  }
  const getListWrapper = getWrapper(List, defaultProps)

  it('renders an unordered list', () => {
    expect(getListWrapper().find('ul').length).toBeTruthy()
  })

  it('renders ListItems', () => {
    expect(getListWrapper().find(ListItem).length).toEqual(items.length)
    expect(
      getListWrapper()
        .find(ListItem)
        .first()
        .text(),
    ).toEqual(items[0].name)
  })

  it('renders an active ListItem', () => {
    expect(
      getListWrapper()
        .find(ListItem)
        .get(defaultProps.activeItemIndex).props.active,
    ).toBe(true)
  })

  it('renders a disabled ListItem', () => {
    const index = 1
    expect(
      getListWrapper()
        .find(ListItem)
        .at(index)
        .props().disabled,
    ).toBe(true)
  })

  it('calls a function when clicked', () => {
    const index = 0
    getListWrapper()
      .find(ListItem)
      .at(index)
      .simulate('click')
    expect(items[index].onClick).toBeCalled()
  })

  everyComponentTestSuite(getListWrapper, List, 'List')
})
