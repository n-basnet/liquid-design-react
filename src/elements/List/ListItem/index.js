import styled, { css } from 'styled-components'

import { ICON_CLASSNAME } from '~/elements/Icon'
import { cursorValue } from '~/utils/styling'

export const ListItem = styled.li`
  display: flex;
  align-items: center;
  max-width: 300px;
  min-height: 50px;
  padding: 10px 15px;
  list-style: none;
  font-size: 14px;
  line-height: 1.75;
  :last-child {
    border-radius: 0 0 6px 6px;
    border: none;
  }
  ${props => css`
    border-bottom: 1px solid ${props.theme.colors.sensitiveGrey.darker};
    color: ${props.theme.colors.richBlack};
    ${cursorValue}
    &:hover {
      ${!props.disabled &&
        props.onClick &&
        css`
          background-color: ${props.grey
    ? props.theme.colors.sensitiveGrey.dark
    : props.theme.colors.white};
          color: ${props.theme.colors.primary.base};
          font-weight: 900;
          transition: ${props.theme.transition};
          .${ICON_CLASSNAME} {
            svg {
              fill: ${props.theme.colors.primary.base};
            }
          }
          ${props.onClick &&
            !props.disabled &&
            css`
              cursor: pointer;
            `};
        `};
    }
    ${props.active &&
      css`
        font-weight: 900;
        color: ${props.theme.colors.primary.base};
      `};
    ${props.grey &&
      css`
        background-color: ${props.theme.colors.sensitiveGrey.base};
      `};
    ${props.disabled &&
      css`
        span,
        svg {
          opacity: 0.3;
        }
      `};
    .${ICON_CLASSNAME} {
      margin-right: 10px;
      svg {
      fill: ${
  props.active
    ? props.theme.colors.primary.base
    : props.theme.colors.richBlack
};
    };
  `};
`
