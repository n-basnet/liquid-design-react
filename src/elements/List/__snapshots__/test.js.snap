// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List matches the snapshot 1`] = `
<ul>
  <ListHead
    grey={true}
  >
    <WithTheme(Icon)
      color="richBlack.base"
      name="star"
      size={16}
    />
  </ListHead>
  <ListItem
    active={true}
    disabled={false}
    grey={true}
    icon="star"
    key="0"
    onClick={[Function]}
  >
    <WithTheme(Icon)
      color="primary.base"
      name="star"
      size={16}
    />
    <span>
      item 1
    </span>
  </ListItem>
  <ListItem
    active={false}
    disabled={false}
    grey={true}
    icon="star"
    key="1"
    onClick={[Function]}
  >
    <WithTheme(Icon)
      color="richBlack.base"
      name="star"
      size={16}
    />
    <span>
      item 2
    </span>
  </ListItem>
  <ListItem
    active={false}
    disabled={true}
    grey={true}
    icon="star"
    key="2"
  >
    <WithTheme(Icon)
      color="richBlack.base"
      name="star"
      size={16}
    />
    <span>
      item 3
    </span>
  </ListItem>
</ul>
`;
